### PR TITLE
fix task local variables execution id replacement

### PR DESCRIPTION
**Pull request by balsarori**
Following discussion in PR #1612

### CODE REVIEW

File: modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java [hunk 1]
```diff
@@ -407,8 +409,12 @@ public void recordExecutionReplacedBy(ExecutionEntity execution, InterpretableEx
         .executionId(execution.getId())
         .unfinished()
         .list();
+      
+      // keep record of unfinished tasks to only update related variables later
+      Set<String> unfinishedTasks = new HashSet<String>();
       for (HistoricTaskInstanceEntity historicTaskInstance: historicTaskInstances) {
         historicTaskInstance.setExecutionId(replacedBy.getId());
+        unfinishedTasks.add(historicTaskInstance.getId());
```
Comment (daisuke-yoshimoto) [inline modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java]: We need to consider unfinished entities as well, so we need to add the cached entities to unfinishedTasks. Specifically, it is the following part.

https://github.com/Activiti/Activiti/blob/a43ff77ed55f881abeb7da87125b29f0bbd30df6/modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java#L403

File: modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java [hunk 2]
```diff
@@ -423,7 +429,11 @@ public void recordExecutionReplacedBy(ExecutionEntity execution, InterpretableEx
         .executionId(execution.getId())
         .list();
       for (HistoricVariableInstanceEntity historicVariableInstance: historicVariableInstances) {
-          historicVariableInstance.setExecutionId(replacedBy.getId());
+    	  // only variables of unfinished tasks
+    	  if (unfinishedTasks.contains(historicVariableInstance.getTaskId())) {
```
Comment (daisuke-yoshimoto) [inline modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java]: We need to consider the cached entities that is not flushed as well, so we need to add this filter to the cached entities as well. Specifically, it is the following part.
https://github.com/Activiti/Activiti/blob/a43ff77ed55f881abeb7da87125b29f0bbd30df6/modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java#L421

File: modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java (hunk 1 above)
Comment (balsarori) [reply modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java]: do you have a case when this is needed?

File: modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java (hunk 2 above)
Comment (balsarori) [reply modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java]: I've also included cached tasks to the check

File: modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java (hunk 1 above)
Comment (daisuke-yoshimoto) [reply modules/activiti-engine/src/main/java/org/activiti/engine/impl/history/DefaultHistoryManager.java]: @balsarori 
I am sorry to have kept you waiting.

I added the case when this is need in the following pull request.
https://github.com/Activiti/Activiti/pull/1739

**Pull request merged**